src/goto-instrument/contracts/contracts.cpp

```diff
@@ -772,35 +769,39 @@ void code_contractst::apply_function_contract(
   is_fresh.add_memory_map_decl(new_program);
 
   // Insert assertion of the precondition immediately before the call site.
-  exprt::operandst requires_conjuncts;
-  for(const auto &r : type.requires())
+  for(const auto &clause : type.requires())
   {
-    requires_conjuncts.push_back(
-      to_lambda_expr(r).application(instantiation_values));
-  }
-  auto requires = conjunction(requires_conjuncts);
-  requires.add_source_location() =
-    requires_conjuncts.empty() ? type.source_location()
-                               : type.requires().front().source_location();
-  if(!requires.is_true())
-  {
-    if(has_subexpr(requires, ID_exists) || has_subexpr(requires, ID_forall))
-      add_quantified_variable(requires, mode);
-
-    goto_programt assertion;
-    converter.goto_convert(code_assertt(requires), assertion, mode);
-    assertion.instructions.back().source_location_nonconst() =
-      requires.source_location();
-    assertion.instructions.back().source_location_nonconst().set_comment(
-      "Check requires clause");
-    assertion.instructions.back().source_location_nonconst().set_property_class(
-      ID_precondition);
-    is_fresh.update_requires(assertion);
-    new_program.destructive_append(assertion);
+    auto instantiated_clause =
+      to_lambda_expr(clause).application(instantiation_values);
+    if(!instantiated_clause.is_true())
+    {
+      if(
+        has_subexpr(instantiated_clause, ID_exists) ||
+        has_subexpr(instantiated_clause, ID_forall))
+        add_quantified_variable(instantiated_clause, mode);
+
+      goto_programt assertion;
+      converter.goto_convert(
+        code_assertt(instantiated_clause), assertion, mode);
+      assertion.instructions.back().source_location_nonconst() =
+        clause.source_location();
+      assertion.instructions.back().source_location_nonconst().set_line(
+        location.get_line());
+      assertion.instructions.back().source_location_nonconst().set_comment(
+        std::string("Check ")
+          .append(target_function.c_str())
+          .append("'s requires clause in ")
+          .append(function.c_str()));
+      assertion.instructions.back()
+        .source_location_nonconst()
+        .set_property_class(ID_precondition);
+      is_fresh.update_requires(assertion);
+      new_program.destructive_append(assertion);
+    }
```

Comment:
1. Could all of this be moved into a function of its own? It's getting way too long.
2. I would advise against the "'s" and use "requires clause of function <name>" to avoid the oddity of using "'s" with a function the name of which ends in "s."

Reply:
I agree on both points.

Reply:
Implemented as suggested. I also refactored the initialization of history variables.

Reply:
Much cleaner now.